### PR TITLE
Fix #5602 Can't add required attribute with default value

### LINKED ISSUE #5602: Can't add required attribute with default value
#### Reproduce
1. Add a new int attribute in the Attribute table:
a. Set *nullable = false*
b. Set *default value = 3*

#### Expected
The attribute is added and all int values are set to 3

#### Observed
Attribute could not be created. The attribute 'newInt' of entity 'YourEntity' contains null values.

### PR DESCRIPTION
**Pull request by dennishendriksen**
Note: integration tests are failing until #6813 has been merged.

#### Checklist
- [ ] Functionality works & meets specifications
- [x] Code reviewed
- [x] Code unit/integration/system tested
- [x] User documentation updated
- (n.a.) (If you have changed REST API interface) view-swagger.ftl updated
- (n.a.) Test plan template updated
- [x] Clean commits

### CODE REVIEW

**Comment by dennishendriksen** [comment]
Integration tests fail due to an exposed issue in the EMX importer.

**Comment by tommydeboer** [comment]
When saving changes in the Application Settings, there's an error:

```Updating read-only attribute 'ga_acc_privacy_friendly_mgs' of entity 'sys_set_app' with id 'app' is not allowed.```

**Comment by tommydeboer** [comment]
It might be a good idea to add an EMX file with some attributes with default values and add that to the list of files that get imported during the importer integration tests.

**Comment by tommydeboer** [comment]
In EMX, you don't have to specify a column in the data sheet when the attribute is nullable. 

However, when having a required attribute with a default value you are required to add the column in the data sheet even when it is empty. Should we consider changing this?

File: molgenis-data-postgresql/src/test/java/org/molgenis/data/postgresql/PostgreSqlQueryGeneratorTest.java [hunk 1]
```diff
@@ -729,7 +775,8 @@ public void getSqlAddColumnInvalidType(AttributeType attrType)
 		EntityType entityType = when(mock(EntityType.class).getId()).thenReturn("entity").getMock();
 		Attribute attr = when(mock(Attribute.class).getName()).thenReturn("attr").getMock();
 		when(attr.getDataType()).thenReturn(attrType);
-		PostgreSqlQueryGenerator.getSqlAddColumn(entityType, attr);
+		PostgreSqlQueryGenerator.getSqlAddColumn(entityType, attr,
+				null); // // FIXME test and make sure we have tests for both modes
```
[the remark this answers is not in the document]
Comment (dennishendriksen) [reply molgenis-data-postgresql/src/test/java/org/molgenis/data/postgresql/PostgreSqlQueryGeneratorTest.java]: Done

File: molgenis-data-postgresql/src/test/java/org/molgenis/data/postgresql/PostgreSqlRepositoryCollectionTest.java [hunk 2]
```diff
@@ -544,8 +544,29 @@ public void addAttribute()
...
+		when(attr.getDataType()).thenReturn(STRING);
+		when(attr.getDefaultValue()).thenReturn("defaultValue");
+		postgreSqlRepoCollection.addAttribute(entityType, attr);
+		verify(jdbcTemplate).execute(
+				"ALTER TABLE \"entity#6844280e\" ADD \"attr\" character varying(255) NOT NULL DEFAULT 'defaultValue'");
+		verify(jdbcTemplate).execute("ALTER TABLE \"entity#6844280e\" ALTER COLUMN \"attr\" DROP DEFAULT");
+		verifyNoMoreInteractions(jdbcTemplate);
 	}
 
+	// TODO add addAttributeDefaultValueMref test once PostgreSqlRepository does not create new JdbcTemplate instances
```
[the remark this answers is not in the document]
Comment (dennishendriksen) [reply molgenis-data-postgresql/src/test/java/org/molgenis/data/postgresql/PostgreSqlRepositoryCollectionTest.java]: https://github.com/molgenis/molgenis/issues/6826
I've removed the TODO

File: molgenis-data/src/main/java/org/molgenis/data/support/AttributeUtils.java [hunk 3]
```diff
@@ -1,15 +1,32 @@
...
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.StreamSupport.stream;
 import static org.molgenis.data.meta.AttributeType.*;
+import static org.molgenis.util.EntityUtils.getTypedValue;
+import static org.molgenis.util.MolgenisDateFormat.*;
 import static org.springframework.util.StringUtils.capitalize;
 
 public class AttributeUtils
 {
```
[the remark this answers is not in the document]
Comment (dennishendriksen) [reply molgenis-data/src/main/java/org/molgenis/data/support/AttributeUtils.java]: Fixed EnumSet --> Set warning
I propose to fix the RuntimeException warnings as part of M5327 Global exception handling
I created https://github.com/molgenis/molgenis/issues/6845 for the getIdAttribute possible NPE warnings

**Comment by dennishendriksen** [comment]
> When saving changes in the Application Settings, there's an error:

Cannot reproduce:
image

**Comment by dennishendriksen** [comment]
> It might be a good idea to add an EMX file with some attributes with default values and add that to the list of files that get imported during the importer integration tests.

I've added a note to the specs story: http://www.molgenis.org/ticket/4602#comment:29

**Pull request merged**